### PR TITLE
Remove addons modules on install and add postinstall execution

### LINKED ISSUE #24638: Remove API push from core, move it into module / lib
#### Is your feature request related to a problem?

This Feature is a part of split between corporate prestashop/Open-sour prestashop.
Goal is to provide a full "non corporate" project.

API push only exists in PS distribution, so it needs to be included in the build pipeline

#### Describe the solution you'd like

A clear and concise description of what you'd want to happen.

### PR DESCRIPTION
**Pull request by PierreRambaud**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Remove addons modules download on install, add the capability to install all modules present on the disk inside the module directory. Also, add a postInstall behavior to be able to execute code after the PrestaShop installation is done. The welcome module has been updated because the module has a bug that break the installation process because of undeclared PS_ADMIN_DIR during the instantiation of the module.
| Type?             | improvement
| Category?         | IN
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | Fixes #24638
| How to test?      | Install PrestaShop, no addons modules are installed.
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

## New behavior & Bc breaks

- Remove addons requests which was downloaded PrestaShop enterprise modules
- Install all modules present inside the `modules` directory
- Add `postInstall` method inside modules
- Remove `getModulesList` and `getAddonsModulesList` inside the `src/PrestaShopBundle/Install/Install.php`
- Rename `getModulesList` into `getModulesOnDisk`
- Remove `processInstallAddonsModules` inside the `install-dev/controllers/console/process.php`
- Remove `processInstallAddonsModules` inside the `install-dev/controllers/http/process.php`
- Interface `PrestaShop\PrestaShop\Core\Addon\Module\ModuleInterface` is now strict typed

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24676)

### CODE REVIEW

File: src/Adapter/Module/Module.php [hunk 1]
```diff
@@ -257,6 +257,18 @@ public function onInstall()
         return $result;
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function onPostInstall(): bool
```
Comment (Progi1984) [inline src/Adapter/Module/Module.php:263]: Could you test it ?

File: src/Core/Addon/Module/ModuleManager.php [hunk 2]
```diff
@@ -317,6 +307,34 @@ public function install($source)
...
 
+
+    /**
+     * Execute post install
+     *
+     * @param string $moduleName
+     *
+     * @return bool true for success
+     */
+    public function postInstall(string $moduleName): bool
```
Comment (Progi1984) [inline src/Core/Addon/Module/ModuleManager.php:325]: Could you test it ?

File: src/PrestaShopBundle/Install/Install.php [hunk 3]
```diff
@@ -892,208 +893,92 @@ public function configureShop(array $data = [])
         return true;
     }
 
-    public function getModulesList()
```
Comment (Progi1984) [inline src/PrestaShopBundle/Install/Install.php:895]: BC Break

File: src/PrestaShopBundle/Install/Install.php [hunk 4]
```diff
@@ -892,208 +893,92 @@ public function configureShop(array $data = [])
...
-            'statsproduct',
-            'statsregistrations',
-            'statssales',
-            'statssearch',
-            'statsstock',
-            'welcome',
-        ];
-    }
-
-    public function getAddonsModulesList($params = [])
```
Comment (Progi1984) [inline src/PrestaShopBundle/Install/Install.php:955]: BC Break

File: src/PrestaShopBundle/Install/Install.php [hunk 5]
```diff
@@ -892,208 +893,92 @@ public function configureShop(array $data = [])
...
-
-        return $addons_modules;
-    }
-
     /**
-     * PROCESS : installModules
-     * Download module from addons and Install all modules in ~/modules/ directory.
+     * Get all modules present on the disk
      */
-    public function installModulesAddons($module = null)
```
Comment (Progi1984) [inline src/PrestaShopBundle/Install/Install.php:1026]: BC Break

File: src/PrestaShopBundle/Install/Install.php [hunk 6]
```diff
@@ -892,208 +893,92 @@ public function configureShop(array $data = [])
...
 
     /**
      * PROCESS : installModules
      * Download module from addons and Install all modules in ~/modules/ directory.
      *
-     * @param array|string|null $module Module to install. If not provided, installs all modules.
-     *
      * @return bool
      */
-    public function installModules($module = null)
```
Comment (Progi1984) [inline src/PrestaShopBundle/Install/Install.php:1051]: BC Break

File: src/PrestaShopBundle/Install/Install.php [hunk 7]
```diff
@@ -892,208 +893,92 @@ public function configureShop(array $data = [])
...
-                continue;
-            }
-
             $moduleException = null;
 
             try {
-                $moduleInstalled = $moduleManager->install($module_name);
-            } catch (\PrestaShopException $e) {
-                $moduleInstalled = false;
+                $moduleEventIsExecuted = $moduleManager->{$event}($module_name);
```
Comment (Progi1984) [inline src/PrestaShopBundle/Install/Install.php]: May be check if method 'pre'.$event ou 'post'.$event exist and execute it

Comment (Progi1984) [reply src/PrestaShopBundle/Install/Install.php]: No hook for event

File: install-dev/controllers/http/process.php [hunk 8]
```diff
@@ -236,21 +236,18 @@ public function processInstallModules()
     }
 
     /**
-     * PROCESS : installModulesAddons
-     * Install modules from addons
+     * PROCESS: Post installation execution
      */
-    public function processInstallAddonsModules()
```
Comment (Progi1984) [inline install-dev/controllers/http/process.php:242]: BC Break

File: install-dev/controllers/console/process.php [hunk 9]
```diff
@@ -289,12 +290,11 @@ public function processInstallFixtures()
     }
 
     /**
-     * PROCESS : installModulesAddons
-     * Install modules from addons
+     * Process post install execution
      */
-    public function processInstallAddonsModules()
```
Comment (Progi1984) [inline install-dev/controllers/console/process.php:295]: BC Break

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/PrestaShopBundle/Install/Install.php (hunk 7 above)
Comment (PierreRambaud) [reply src/PrestaShopBundle/Install/Install.php]: What do you mean by "no hook for event"? Please use a proper sentence to be understandable :smile:

File: src/Adapter/Module/Module.php (hunk 1 above)
Comment (PierreRambaud) [reply src/Adapter/Module/Module.php:263]: I can't test it, this class rely on Object model with `LegacyModule` :(

File: src/Core/Addon/Module/ModuleManager.php (hunk 2 above)
Comment (PierreRambaud) [reply src/Core/Addon/Module/ModuleManager.php:325]: Same I can't test it, too much dependency from the database and fs.
The current state of these files are not designed for unit testing :sob:

File: src/PrestaShopBundle/Install/Install.php (hunk 7 above)
Comment (PierreRambaud) [reply src/PrestaShopBundle/Install/Install.php]: We don't need to check if the method exists as we control the call, and this is not a public method.

**Comment by atomiix** [comment]
I think you forgot to remove "addons_modules" from [accepted steps](https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/controllers/console/process.php#L106) and add "postInstall"

File: tests/Unit/Core/Addon/Module/ModuleManagerTest.php [hunk 10]
```diff
@@ -78,6 +78,12 @@ public function testInstallSuccessful()
         $this->assertTrue($this->moduleManager->install(self::INSTALLED_MODULE));
     }
 
+    public function testPostInstallSuccessful()
```
Comment (Progi1984) [inline tests/Unit/Core/Addon/Module/ModuleManagerTest.php]:
```suggestion
    public function testPostInstallSuccessful(): void
```

**Review by Progi1984 — COMMENTED** [review]
A last feedback (but it's approved)

Ping @boubkerbribri for reviewing santity tests ;)

**Review by atomiix — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

File: src/PrestaShopBundle/Install/Install.php [hunk 11]
```diff
@@ -892,208 +893,92 @@ public function configureShop(array $data = [])
...
-        return $addons_modules;
-    }
-
     /**
-     * PROCESS : installModules
-     * Download module from addons and Install all modules in ~/modules/ directory.
+     * Get all modules present on the disk
      */
-    public function installModulesAddons($module = null)
+    public function getModulesList(): array
```
Comment (jolelievre) [inline src/PrestaShopBundle/Install/Install.php]:
```suggestion
    public function getModulesOnDisk(): array
```

File: src/PrestaShopBundle/Install/Install.php [hunk 12]
```diff
@@ -892,208 +893,92 @@ public function configureShop(array $data = [])
...
+    public function postInstall(): bool
+    {
+        return $this->executeEvent(
+            $this->getModulesList(),
+            'postInstall',
+            'Cannot execute post install on module "%module%"'
+        );
+    }
+
+    protected function executeEvent(array $modules, string $event, string $errorMessage): bool
```
Comment (jolelievre) [inline src/PrestaShopBundle/Install/Install.php]:
```suggestion
    protected function executeAction(array $modules, string $action, string $errorMessage): bool
```

File: src/PrestaShopBundle/Install/Install.php [hunk 13]
```diff
@@ -892,208 +893,92 @@ public function configureShop(array $data = [])
...
-
             $moduleException = null;
 
             try {
-                $moduleInstalled = $moduleManager->install($module_name);
-            } catch (\PrestaShopException $e) {
-                $moduleInstalled = false;
+                $moduleEventIsExecuted = $moduleManager->{$event}($module_name);
+            } catch (PrestaShopException $e) {
+                $moduleEventIsExecuted = false;
```
Comment (jolelievre) [inline src/PrestaShopBundle/Install/Install.php]:
```suggestion
                $moduleActionIsExecuted = false;
```

File: src/Core/Addon/Module/ModuleInterface.php [hunk 14]
```diff
@@ -32,6 +32,8 @@ interface ModuleInterface extends AddonInterface
 {
     public function onInstall();
```
Comment (jolelievre) [inline src/Core/Addon/Module/ModuleInterface.php]: Maybe you can strong type all the other functions, since there is already a BC in this interface

**Review by jolelievre — COMMENTED** [review]
Just a few naming suggestions

**Review by boubkerbribri — DISMISSED** [review]
Tests UI OK

**Review by Progi1984 — DISMISSED** [review]

**Review by boubkerbribri — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Comment by Progi1984** [comment]
@Julievrz If you have a little time :)

File: src/PrestaShopBundle/Install/Install.php [hunk 15]
```diff
@@ -892,208 +893,92 @@ public function configureShop(array $data = [])
...
                 $moduleException = $e->getMessage();
             }
 
-            if (!$moduleInstalled) {
-                $module_errors = [$this->translator->trans('Cannot install module "%module%"', ['%module%' => $module_name], 'Install')];
+            if (!$moduleActionIsExecuted) {
+                $moduleErrors = [
+                    $this->translator->trans(
+                        $errorMessage,
+                        ['%module%' => $module_name],
```
Comment (Julievrz) [inline src/PrestaShopBundle/Install/Install.php]: Just have a question regarding this string, did you remove the rest of the sentence on purpose? :)

**Review by Julievrz — COMMENTED** [review]
The new wording is ok, I just have one question. Thanks

**Review by jolelievre — DISMISSED** [review]

File: src/PrestaShopBundle/Install/Install.php (hunk 15 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Install/Install.php]: If you use a variable like this we won't be able to scan the wordings (especially the new one for post install)

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by Julievrz — DISMISSED** [review]
Wording ✔️ Thanks!

**Review by Progi1984 — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Comment by Robin-Fischer-PS** [comment]
HI @PierreRambaud  !

I tested :

 * Installation : OK
 * Check the addons Modules are no longer installed by default : OK
 * Log in on addons in BO
 * Log out of addons in BO
 * Installation with a new module (ps_qualityassurance) added : Automatic install of the module OK.
 * Installation of module via Module Catalog : OK
 * Installation of module via zip upload : OK
 * Upgrade of module via module manager : OK
 * Shop still OK (create order, customer, etc...)

So it's QA :heavy_check_mark: 

Thanks :)

**Pull request merged**